### PR TITLE
Add theme support for AMP compatibility with official plugin

### LINKED ISSUE #77: Implement AMP template and style
The AMP format seems to be constantly gaining traction and it would be interesting to offer it as part of the parent theme.

### PR DESCRIPTION
**Pull request by westonruter**
Fixes #77.

This adds `amp` theme support for the [official AMP plugin](https://wordpress.org/plugins/amp). The existing theme templates and styles are used, but places where JS was used are replaced with the AMP equivalents.

* Add `amp` theme support with `paired` flag to indicate theme works in both Standard (AMP-first) and Transitional modes (paired AMP).
* Add `primer_is_amp()` helper function.
* Prevent enqueueing scripts that are not compatible with AMP.
* Implement AMP interactions for toggling mobile menu. This replaces the `Primer_Walker_Nav_Menu` with a function that filters `walker_nav_menu_start_el` to inject the submenu toggles. I opted to not use the [`nav_menu_dropdown`](https://amp-wp.org/documentation/playbooks/navigation-sub-menu-buttons/) and [`nav_menu_toggle`](https://amp-wp.org/documentation/playbooks/toggling-hamburger-menus/) theme support flags in favor of just making the changes directly in the theme, as it was actually easier to do so.
* Add accessibility to nav menus when JS is disabled or in AMP; improve semantics/ARIA of elements.
* Remove apparently unused JS code for nav menus.
* Remove external `skip-link-focus-fix` script in favor of inlining the logic, following what was done in Twenty Nineteen. 

The remaining item to do is to fix the header video when YouTube is used, but this depends on an upstream change to AMP itself: https://github.com/ampproject/amp-wp/pull/2642.

### CODE REVIEW

**Comment by westonruter** [comment]
In regards to the video header, I tried using an MP4 on the non-AMP version and it seemed broken:

image

The video file ([VID_20190510_140903.mp4.zip](https://github.com/godaddy/wp-primer-theme/files/3293932/VID_20190510_140903.mp4.zip)) looks like this:

image

So I was expecting to see the bridge, not just the top-left corner of the sky. I also didn't expect to see a black margin on the left. So in a6104c3 I fixed what appeared to be a bug with the layout a header `video` element, and added AMP compat.

# Non-AMP

image

# AMP

image

----

Aside: This also unveiled a bug with the AMP plugin's general handling of video headers, in that the `loop` attribute was failing to be added:

```diff
--- a/includes/class-amp-theme-support.php
+++ b/includes/class-amp-theme-support.php
@@ -2408,7 +2408,8 @@ class AMP_Theme_Support {
 				array_merge(
 					$video_attributes,
 					array(
-						'src' => $video_settings['videoUrl'],
+						'src'  => $video_settings['videoUrl'],
+						'loop' => '',
 					)
 				)
 			);
```

I also noticed another issue with the AMP plugin where if you do not supply a header image at all, then a selected video does not render. For AMP fixes to these issues, see https://github.com/ampproject/amp-wp/pull/2642.

File: assets/js/navigation.js [hunk 1]
```diff
@@ -3,13 +3,13 @@
...
-	    $menuToggle = false,
-	    $submenu    = false;
+	    $menuToggle = false;
 
 	function toggle() {
 
 		$menuToggle.add( $navMenu ).toggleClass( 'open' );
 
+		// @todo Can this be removed? It seems somewhat annoying that the expanded submenus would reset when toggling the menu.
 		resetSubmenuToggles();
```
Comment (EvanHerman) [inline assets/js/navigation.js]: That's a good point. I can see how one would expect the submenus to remain expanded when toggling the menus. Let's get rid of this.

File: functions.php [hunk 2]
```diff
@@ -321,6 +314,26 @@ function primer_setup() {
...
+	 *
+	 * This is only done for version v1.2 and greater because in this version support was added for CSS tree shaking
+	 * and the AMP toggleClass action. Also, this is the version that adds the ability to switch the mode even when
+	 * the theme support flag is present.
+	 *
+	 * @link  https://amp-wp.org
+	 * @since 1.9
+	 */
+	if ( defined( 'AMP__VERSION' ) && version_compare( strtok( AMP__VERSION, '-' ), '1.2', '>=' ) ) {
+		add_theme_support(
```
Comment (EvanHerman) [inline functions.php:329]: Can we add empty spaces before/after `add_theme_support`?

```php
if ( defined( 'AMP__VERSION' ) && version_compare( strtok( AMP__VERSION, '-' ), '1.2', '>=' ) ) {

   add_theme_support(...

}
```

File: functions.php [hunk 3]
```diff
@@ -511,10 +524,12 @@ function primer_scripts() {
 
 	$nav_dependencies = ( is_front_page() && function_exists( 'has_header_video' ) && has_header_video() ) ? array( 'jquery', 'wp-custom-header' ) : array( 'jquery' );
 
-	wp_enqueue_script( 'primer-navigation', get_template_directory_uri() . "/assets/js/navigation{$suffix}.js", $nav_dependencies, PRIMER_VERSION, true );
-	wp_enqueue_script( 'primer-skip-link-focus-fix', get_template_directory_uri() . "/assets/js/skip-link-focus-fix{$suffix}.js", array(), PRIMER_VERSION, true );
+	// The interactivity of the menu in AMP is defined inline.
+	if ( ! primer_is_amp() ) {
+		wp_enqueue_script( 'primer-navigation', get_template_directory_uri() . "/assets/js/navigation{$suffix}.js", $nav_dependencies, PRIMER_VERSION, true );
```
Comment (EvanHerman) [inline functions.php:532]: Can we add an empty line before & after this line?

```php
if ( ! primer_is_amp() ) {
		
   wp_enqueue_script( 'primer-navigation', get_template_directory_uri() . "/assets/js/navigation{$suffix}.js", $nav_dependencies, PRIMER_VERSION, true );

}

File: functions.php [hunk 4]
```diff
@@ -535,6 +550,32 @@ function primer_scripts() {
...
+ * This does not enqueue the script because it is tiny and because it is only for IE11,
+ * thus it does not warrant having an entire dedicated blocking script being loaded.
+ * This minified code comes from the Twenty Nineteen theme; refer to the theme for the
+ * unminified source. Since IE11 marketshare is very small, this code can likely be
+ * eliminated entirely soon.
+ *
+ * @see twentynineteen_skip_link_focus_fix()
+ * @link https://git.io/vWdr2
+ */
+function primer_skip_link_focus_fix() {
```
Comment (EvanHerman) [inline functions.php:569]: Proper spacing here (Add empty lines to pad code. Eases in readability)

File: header.php [hunk 5]
```diff
@@ -52,9 +52,13 @@
 		 */
 		do_action( 'primer_before_header' );
 
+		$masthead_classes = array( 'site-header' );
+		if ( has_header_video() && primer_is_amp() ) {
```
Comment (EvanHerman) [inline header.php:57]: Proper spacing here.

```php
$masthead_classes = array( 'site-header' );

if ( has_header_video() && primer_is_amp() ) {
```

etc.

File: inc/helpers.php [hunk 6]
```diff
@@ -133,6 +133,18 @@ function primer_is_fixed_width() {
...
+/**
+ * Check if the response is AMP.
+ *
+ * @since 1.9
+ * @link https://wordpress.org/plugins/amp/
+ *
+ * @return bool If an AMP response.
+ */
+function primer_is_amp() {
+	return function_exists( 'is_amp_endpoint' ) && is_amp_endpoint();
```
Comment (EvanHerman) [inline inc/helpers.php:146]: Add empty line before/after.

File: inc/hooks.php [hunk 7]
```diff
@@ -131,16 +131,57 @@ function primer_add_primary_menu() {
 
 	}
 
+	add_filter( 'walker_nav_menu_start_el', 'primer_add_primary_nav_sub_menu_buttons', 10, 3 );
```
Comment (EvanHerman) [inline inc/hooks.php:134]: Spacing

File: inc/hooks.php [hunk 8]
```diff
@@ -131,16 +131,57 @@ function primer_add_primary_menu() {
...
 	wp_nav_menu(
 		array(
 			'theme_location' => 'primary',
-			'walker'         => new Primer_Walker_Nav_Menu,
 		)
 	);
-
+	remove_filter( 'walker_nav_menu_start_el', 'primer_add_primary_nav_sub_menu_buttons', 10 );
```
Comment (EvanHerman) [inline inc/hooks.php:142]: Spacing

**Review by EvanHerman — CHANGES_REQUESTED** [review]
@westonruter Everything looks great here and works well. Thank you for taking the time to put this together and issue a PR.

Can we just add some spacing (empty lines) to some of the lines of code? This will just help to maintain code styles throughout Primer.

File: assets/js/navigation.js (hunk 1 above)
Comment (westonruter) [reply assets/js/navigation.js]: Removed in 4d36f81

**Review by EvanHerman — APPROVED** [review]
Thank you @westonruter! Appreciate all the hard work on this 👍

**Comment by westonruter** [comment]
Great. When should this be released on WordPress.org?

**Comment by EvanHerman** [comment]
@westonruter 

This was rolled into the 1.8.7 release which has been available on WordPress.org for the last two hours.

**Pull request merged**